lms/views/email.py

```diff
@@ -55,33 +55,23 @@ def unsubscribe(self):
 
         return HTTPFound(location=self.request.route_url("email.unsubscribed"))
 
-    @view_config(
-        route_name="email.preferences",
-        request_method="GET",
-        request_param="token",
-    )
-    def preferences_redirect(self):
-        # The token has already been verified by the security policy, so we can
-        # just go right ahead with the redirect.
-        return HTTPFound(
-            location=self.request.route_url("email.preferences"),
-            # Set a cookie to keep the user logged in even though we're
-            # removing the authentication token from the URL.
-            headers=remember(self.request, self.request.authenticated_userid),
-        )
-
     @view_config(
         route_name="email.preferences",
         request_method="GET",
         renderer="lms:templates/email/preferences.html.jinja2",
     )
     def preferences(self):
+        days = self.email_preferences_service.get_preferences(
+            self.request.authenticated_userid
+        ).days()
+        self.request.response.headers = remember(
+            self.request, self.request.authenticated_userid
+        )
```

Comment:
The `preferences()` view now needs to set the cookie, which was previously being set by the `preferences_redirect()` view.

This line will get executed and will set the cookie header with a newly-fresh cookie every time you reload the preferences page (or submit the form) but that's fine and in fact Pyramid was already doing so automatically anyway. It's not necessary to execute this line every time but it does no harm and it's simpler than trying to do it only once.